doc/design/upgrade.md

```diff
@@ -0,0 +1,118 @@
+
+In the future, will consider to add live migration, so user does not experience service broken during upgrade. Integration docker log with syslogd is another option to reduce impact of VCH downtime.
+
+### VCH status
+
+vic-machine ls and inspect should be able to show the VCH versions and upgrade status.
+
+### Roll Back
+
+For anything wrong happens during upgrade, vic-machine will rollback any changes made to have VCH return to original version and status. If VC/vSphere connection is broken during upgrade, the upgrade or rollback will fail, but user should be able to rollback previously broken upgrade, to recover VCH.
+
+### Downgrade
+
+For any running VCH, VIC admin can downgrade existing VCH to an old version. (I assume this function is only used for upgrade roll back, we might not want to recommend downgrade for anything working well, then the downgrade scope will be same to upgrade versions)
+
+Verified docker-engine upgrade/downgrade, which works well between 1.11.2 and 1.12. The running container will be stopped after upgrade/downgrade, but is good to start again in new version. And all images and containers information is not lost. So from technical point of view, we should support same story.
+
+### Internet connection
+Internet connection is not required to upgrade/downgrade VCH, but newer version's binary should be available for vic-machine.
+
+## Implementation/Issues
+
+### Versioning
+
+#### Freeze following attributes definition in VCH configuration and container VM configuration
+- VCH configuration: ID, Name, Version
+- Container VM serial port configuration.
+- Container VM log/debug file location.
+
+#### Configuration Version
+
+Both VCH and container configuration will have version and the value is same with what vic-machine version command shown. After upgrade, VCH configuration version should be updated to new version, but container VM configuration version will still be old one.
+
+New version's VCH will work with both old and new versions containers.
+
+#### Embed iso file version
+
+vic-machine need to identify the iso file version before and after upgrade. Two options here:
+- Add version into iso file Primary Volume Descriptor, and read back from vic-machine
+- Leverage iso file name including version, e.g. appliance-0.5.0.iso
+
+vic-machine should check iso file version during deployment, and after upload to datastore, version should be appended to iso file name, to make sure mutilple iso files could co-exist in the same VCH, and used for different version's container.
+
+Image file name should not be changed in datastore, cause the file path is used to create container VM. And then vic-machine will leverage file name for version checking during upgrade, to avoid download iso file from datastore back to where vic-machine is running.
+
+Note: No feasible golang library found for this function, so will write our own library to read iso metadata.
```

Comment:
The iso9660 is to extract iso files, not read iso file itself. 
The iso9660 standard defines where to put some metadata, and we've used the publisher field for our information. So I reused that for vic version information. But no go library found to read the field out.
But it's easy to do that ourselves.